Implementation Plan: Nuxt 3 migration

obulat opened the pull request:
<!-- prettier-ignore -->
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->

Resolves #2324 by @obulat

## Reviewers
- [ ] @sarayourfriend - for experience with the previous attempts at migrating to Nuxt 3
- [ ] @zackkrida - for experience with initial conversion from Vue SSR to Nuxt

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
This is a plan of how I propose to approach Nuxt 3 migration. I am especially curious about ideas on "Nuxt 3 / Vue 3 features in Openverse: discussion" section.  It must be noted that there will be changes to this plan during its implementation due to the unknowns of this process.

Rendered plan: https://docs.openverse.org/_preview/2350/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.html

This discussion is following the Openverse decision-making process. Information about this process can be found [on the Openverse documentation site](https://docs.openverse.org/meta/decision_making/index.html). Requested reviewers or participants will be following this process. If you are being asked to give input on a specific detail, you do not need to familiarise yourself with the process and follow it.

## Current round

This discussion is currently in the Clarification round.

The deadline for review of this round is TBD.

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

obulat commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,278 @@
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @sarayourfriend - for extensive experience with JS
+- [ ] @zackkrida - for extensive frontend experience and the previous migration
+      to Nuxt
```
```suggestion
- [ ] @sarayourfriend
- [ ] @zackkrida
```

sarayourfriend commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+| `"@nuxtjs/svg-sprite"`             | Update to current that supports Nuxt 3                                                               |                                                                                                                                                          |
+| `"@nuxtjs/eslint-module"`          | Update to current that supports Nuxt 3                                                               | https://github.com/nuxt-modules/eslint                                                                                                                   |
+| `"@pinia/nuxt"`                    | Update to current version that supports Nuxt 3                                                       |                                                                                                                                                          |
+| `"@nuxtjs/proxy"`                  | Replace with Nitro routeRules or use `proxyRequest`                                                  | https://stackoverflow.com/questions/76120894/how-do-i-set-proxy-with-baseurl-with-nuxt3-and-nitro-config, https://github.com/nuxt/nuxt/discussions/15907 |
+| `"@nuxtjs/redirect-module"`        | Define Nitro routeRules in `nuxt.config`                                                             | https://nitro.unjs.io/config/#routerules                                                                                                                 |
+| `"@nuxtjs/sentry"`                 | No official support                                                                                  | Workaround in https://github.com/nuxt-community/sentry-module/issues/530                                                                                 |
+| `"vue-plausible"`                  | Update to current version that supports Nuxt 3 or https://github.com/nuxt-modules/plausible          | Note: enabling outbound tracking breaks links with target blank in `vue-plausible`                                                                       |
```
Would each of these that need workarounds be separate issues? I've only looked at the proxy one but it looks certainly complex enough to be its own issue. It also looks like we could implement the workaround for it in Nuxt 2, FWIW, perhaps in conjunction with #2392 which would complicate the current configuration anyway.

obulat replied:
Yes, separating the workarounds into separate issues is a good idea. Adding modules can also be worked on in parallel.

sarayourfriend commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+- [ ] `Vue.set` is not necessary anymore to keep array/object reactivity
+- [ ] [async components change of syntax](https://v3-migration.vuejs.org/breaking-changes/async-components.html)
+      (for components that use `Comp = () => import("component.vue")`)
+- [ ] [`emits` option](https://v3-migration.vuejs.org/breaking-changes/emits-option.html) -
+      replace the `defineEvents` custom implementation
```
These changes do not replace `defineEvent`, as far as I can tell. The `emits` feature described in the link is actually required for `defineEvent` to work in the first place, the docstring for the function specifies that it's a workaround for TS in Composition API: https://github.com//WordPress/openverse/blob/697f62f01a32cb7fcf2f4a7627650a113cba40da/frontend/src/types/emits.ts#L9-L18

I'm surprised that Vue doesn't have a built-in solution for this though, it's very silly to need to write an empty validation function (maybe one that even gets included in your final bundle, I'm not sure if Nuxt shakes those out in build) just to get types for events.

In any case, the `defineEvent` utility was already developed for the feature linked on the documentation page and I don't see anything there or on the Vue 3 `emits` documentation for ways to define the event types without the strange empty validator function workaround.

obulat replied:
I was actually thinking of `defineEmits`: https://vuejs.org/api/sfc-script-setup.html#defineprops-defineemits, more on typing emits: https://vuejs.org/guide/typescript/composition-api.html#typing-component-emits

However, they are only available in components that use `script setup`. 

I think the best course here would be to keep using `defineEvent` with `emits`, and refactor the components to `script setup` and `defineEmits` later, after this project is done. It could be a separate milestone for converting all of the components that emit events to `script setup`.

sarayourfriend replied:
> I think the best course here would be to keep using defineEvent with emits, and refactor the components to script setup and defineEmits later, after this project is done. It could be a separate milestone for converting all of the components that emit events to script setup.

Sounds good to me! I'm now remembering that I wrote `defineEvent` in frustration that `defineEmits` didn't work in traditional components :sweat_smile:

sarayourfriend commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+      auto-import `scss` variables into all components. We don't use `scss`
+      anymore
+
+### Changes in Nuxt 3 branch
```
We need a step in this to remove the `nuxt-templates-overrides` too, right?

obulat replied:
Right!

sarayourfriend commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+- [ ] [`$attrs` includes `class` & `style`](https://v3-migration.vuejs.org/breaking-changes/attrs-includes-class-style.html) -
+      remove `v-bind="$attrs"` from the root element in components where
+      `inheritAttrs="false"`
+- [ ] `Vue.set` is not necessary anymore to keep array/object reactivity
```
Is the current usage of `Vue.set` (only in the feature flags store, as far as I see) able to be removed in Nuxt 2 if we changed the `Vue.set(flag, 'preferredState', ...)` to `this.flags[flagName].preferredState = ...`, etc.?

obulat replied:
Vue 2 cannot detect property addition or deletion in objects. In the feature flags store, we add a "preferredState" property to the object loaded from JSON. So, to make it reactive, we have to use `Vue.set`[^1] in Vue 2 (and Nuxt 2, respectively).

`this.flags[flagName].preferredState` - would be adding a new property to `this.flags[flagName]`, so this new property would not be reactive.

[^1]: https://v2.vuejs.org/v2/guide/reactivity.html#For-Objects

sarayourfriend commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+- [ ] [`$listeners` has been removed / merged into `$attrs`](https://v3-migration.vuejs.org/breaking-changes/listeners-removed)-
+      remove the `$listeners` everywhere and make sure that we have
+      `v-bind="$attrs"` everywhere we had `$listeners` before
+- [ ] [`$attrs` includes `class` & `style`](https://v3-migration.vuejs.org/breaking-changes/attrs-includes-class-style.html) -
+      remove `v-bind="$attrs"` from the root element in components where
+      `inheritAttrs="false"`
```
Can you give more details for why this is the modification required to handle this breaking change? I'm not making the connection between `inheritAttrs="false"` and the class and style attributes being included in `$attrs`.

sarayourfriend commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+  - Move the functions that are called at the app set up from
+    `middleware`/`layout`s to `app.vue`.
+  - Convert the layouts and pages to use `NuxtPage` instead of `NuxtChild` and
+    `slot` instead of `Nuxt`. For other necessary changes, look up the Nuxt 3
+    docs.
```
Would it be possible to describe the necessary Nuxt 3 changes here the way the Vue 3 changes are described?

sarayourfriend reviewed (commented):
This looks good to me so far, and it gives a decent sense of how much we have to do for this. I've left a handful of questions for clarification, but I would also like to request a clearer/more explicit list of the required changes, ordered as though could be converted to issues and perhaps with links to individual sections for ones that need more explanation. Splitting the changes between the modules section, the Vue 3 breaking changes section, and the implementation stage sections makes it hard to get a general overview of how the implementation would go. I appreciate the two implementation stages being explicit, though. How much benefit is there to pushing as much as we can forward into the pre-`nuxt3` branch phase? Seems like it would prevent a lot more merge conflicts, right? Also, should we put a freeze on frontend feature development during the Nuxt 3 conversion phase, with all frontend development focused exclusively on Nuxt 3 conversion? It would be frustrating to have the Nuxt 3 branch almost perfectly working only to run into big feature related merge conflicts (like #2118  and #2126, which haven't started implementation yet).

obulat commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+
+### Nuxt 3 / Vue 3 features in Openverse: discussion
+
+#### Autoimports
+
+Nuxt 3 auto imports composables and components. Unlike the usual global imports,
+these imports are handled in such a way that only the imports that are actually
+used are left in the production files. The Nuxt team have optimized the way they
+work both in the working app and in the IDEs, so I think we _should_ use the
+auto import feature and not go against the framework and disable it. In the
+current setup, our test suite throws warnings when a child component is not
+imported. We would need to make sure that this does not happen with Nuxt 3.
```
I anticipated a similar response :) Thank you for describing the reasons in such a detail, it's really helpful.

On the first point, I looked at the Nuxt 3 draft branch locally to see if the IDE support helps with it. And even though I can click "Go to definition" on a non-imported component and open the component file, I can't do the same for a composable. Going to definition there sends you to the file that lists all of the auto-imported items, so while it _is_ possible to find the source of the composable, it's not as straightforward as when we import it. 
> Variable name collision and shadowing is also much harder to understand or anticipate side effects for.
This is also a serious consideration.

The second point - I guess you can't understand the extent of how bad it can be until that happens to you :)
I agree with the third point, too.

What I wanted to avoid was going against the framework. This would cause us to have to add work-arounds around the magic, which might need a lot of work. However, seeing that the IDE support is not as great as I was expecting, and considering all of the drawbacks of using auto-imports, I'm open to _not_ using this feature.

obulat replied:
After writing this comment, I decided to check if we would have problems with the imports and tests. In the current Nuxt 2 app, the components are auto-imported, so the app can run even if we don't add all component imports. However, the tests show warnings for components that weren't imported.

In my first try, this was a big problem with explicitly adding all of the imports. The biggest problem is the imports from `.nuxt` and the imports that start with `#`, especially deep within modules (`nuxtjs/svg-sprite` being the one I tried to work around the most). Not saying that this is a blocker, but just noting here that getting all of the imports to work both in the prod app and the tests would require a lot of debugging.

sarayourfriend replied:
> And even though I can click "Go to definition" on a non-imported component and open the component file, I can't do the same for a composable. Going to definition there sends you to the file that lists all of the auto-imported items, so while it is possible to find the source of the composable, it's not as straightforward as when we import it.

This reminds me actually, that Nuxt 3's auto-imports I believe is the main reason why running TypeScript _requires_ a build. Nuxt 3's build process pulls the auto-imported composables out and generates the tsconfig.json to include all the ambient types that you're meant to extend in your Nuxt app's tsconfig.json.

I wonder how much all those ambient types slow down TypeScript (or perhaps increase its memory footprint).

In either case, we'd need to change CI to run the types check only after a Nuxt 3 build is run, at least if that process is still the same, which sounds to be the case based on the IDE behaviour of go-to-definition for the composables.

> Not saying that this is a blocker, but just noting here that getting all of the imports to work both in the prod app and the tests would require a lot of debugging

That's a shame :slightly_frowning_face:. When I tried to convert Nuxt using bridge, the option to disable auto-imports didn't even work. Does turning off auto-imports stop component auto-importing in the app build? The Nuxt 3 only reference auto-importing for _composables_, both in the composables documentation and in the auto-imports documentation :thinking:

obulat replied:
> In either case, we'd need to change CI to run the types check only after a Nuxt 3 build is run, at least if that process is still the same, which sounds to be the case based on the IDE behaviour of go-to-definition for the composables.

I will add this to the proposal, thanks! (Check if we need to change CI to run the types check after a Nuxt 3 build is run - change if necessary)

> The Nuxt 3 only reference auto-importing for composables, both in the composables documentation and in the auto-imports documentation 🤔
I'm not sure if I understand this sentence correctly, but Nuxt 3 docs do say that they auto import components, composables and even utils (although the way it's worded is not very straightforward): https://nuxt.com/docs/guide/concepts/auto-imports#directory-based-auto-imports

By the way, I didn't realize that the recommended way to explicitly add imports is by using `#imports`, found while re-reading the docs for this comment.

sarayourfriend replied:
> Check if we need to change CI to run the types check after a Nuxt 3 build is run - change if necessary

This would also need to change in the pre-commit, which is probably the place it will hurt the DevEx the most.

zackkrida replied:
I definitely understand the desire to do what's idiomatic for the framework _and_ absolutely hating this automagical import stuff. 

> Does turning off auto-imports stop component auto-importing in the app build? 

I just tested this with a fresh Nuxt 3 install. Sadly, components still appear to be auto imported even with this config:

```ts
export default defineNuxtConfig({
  imports: {
    autoImport: false
  }
})
```

It _does_ work for composeables, but not components. 

I'm really conflicted on how to proceed for this one.

sarayourfriend replied:
Actually, if it auto-imports components, I think that's fine. That's the current behaviour of Nuxt 2, so it would present effectively zero change to our workflow.

zackkrida replied:
Oh, true!

obulat commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+component. The decision wether to convert components to `script setup` would be
+at the discretion of the PR implementer.
+
+#### Testing
+
+Vue 3 recommends to use Vitest for better performance and integration with
+`vite` (which will replace `webpack` in our setup). I have not had an
+opportunity to test it out, but I think it would make sense to convert our unit
+tests to `Vitest`. The decision on whether to keep using the current
+dependencies or convert to `Vitest` should be based on the ease of conversion.
```
I've found [`eslint-plugin-vitest`](https://github.com/veritem/eslint-plugin-vitest#credits) which says that most of its rules are essentially the ports of `eslint-plugin-jest` with necessary modifications.
I'll add it here.

sarayourfriend replied:
That's fantastic :rocket:

zackkrida replied:
Vitest maintains a migration doc that might be useful to link here:

https://vitest.dev/guide/migration.html#migrating-from-jest

obulat commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+  - Convert the layouts and pages to use `NuxtPage` instead of `NuxtChild` and
+    `slot` instead of `Nuxt`. For other necessary changes, look up the Nuxt 3
+    docs.
+  - Use the new way of declaring `async components`
```
The banners inside the `VBanner`, and headers in the `search-layout`.
https://github.com/WordPress/openverse/blob/41b0cdb37d616b5e53ee45f77f0448b07747e04f/frontend/src/components/VBanner/VBanners.vue#L34-L38

https://github.com/WordPress/openverse/blob/41b0cdb37d616b5e53ee45f77f0448b07747e04f/frontend/src/layouts/search-layout.vue#L73-L75
I'll add this to the plan

obulat commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+  the import and the functionality that uses it.
+- Remove all `v-on="$listeners"`, making sure that the `on` handlers are still
+  passed to the correct component/element.
+- Remove the `.native` modifiers, and convert `VLink` to use `NuxtLink`. The
```
New `NuxtLink` is basically what our `VLink` is: it handles both internal and external links: https://nuxt.com/docs/api/components/nuxt-link , so we could actually replace `VLink` with `NuxtLink`.

When trying to migrate to Nuxt 3 locally, I had some problems with the buttons that use `as="VLink"`, they wouldn't navigate correctly. I'm not sure if it's the problem with `NuxtLink` or the overall pages setup, though.

sarayourfriend replied:
Ahhhh, I see, I didn't realise there was a difference between Nuxt 3 `NuxtLink` and was confused because we already used `NuxtLink` :sweat_smile: The difference might be worth clarifying, especially "so we could actually replace VLink with NuxtLink", which is awesome.

obulat commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,277 @@
+
+<!-- Describe any infrastructure that will need to be provisioned or modified. In particular, identify associated potential cost changes. -->
+
+There should be no changes to infrastructure.
```
I don't really know.

Here are two things I can say when looking through the Docker file:
The `static` will need to be changed to `/` here:
```
RUN echo "{\"release\":\"${SEMANTIC_VERSION}\"}" > /home/node/frontend/src/static/version.json
```

PORT should be set correctly for the production build (I'm not yet sure what exactly "correctly" means here)

I'll add a sentence "Test that the Docker build works as expected".

zackkrida commented:
As I shared with Olga, I'll review this today.

obulat commented:
I've updated the plan to add the suggestions based on @sarayourfriend's comments. I removed the "Breaking changes" section for Vue and Nuxt. Instead, I added the steps of what we need to update with the note if it's a "Vue 3 breaking change". I also tried to add links to documentation to each step.
Some items will need to be edited for clarity, I'll try to do it tomorrow.

sarayourfriend commented:
@zackkrida Will you be able to review this still or are you planning to wait until after the revisions?

zackkrida commented:
I'll be reviewing this prior to any additional revisions. I did want until the _previous_ revisions were made, after chatting about them with Olga.

zackkrida commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,338 @@
+
+- Deprecation of CJS modules (the ones that use `require` and `module.exports`):
+
+  - [ ] `case` package that we use for `kebab-case`, `title` case and
+        `camelCase` function is CJS. We can either find replacements that use
+        `ESM` modules, vendor in these functions, or use a workaround that
+        `vite` suggests (instead of `const { kebab } = require "case"` use
+        `import casePkg; const { kebab } = casePkg;`)
```
I think we could just vendor this and recommend that, rather than listing multiple potential options here.

zackkrida commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,338 @@
+
+We should probably add [`nuxt-vitest`](https://github.com/danielroe/nuxt-vitest)
+for testing in Nuxt. The package is under active development, so we should pin
+to the patch version
```
```suggestion
to the patch version.
```

zackkrida commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,338 @@
+      ClientOnly, so the current implementation is causing server-client
+      mismatch and rehydration. We should move to a simple HTML anchor tag and
+      an `id` on the content.
+- [ ] Refactor the functionality that is called on app load and route
+      navigation.
```
Could you elaborate on what this entails? It feels really vague and I'm not sure what it refers to.

zackkrida commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,338 @@
+
+## Blockers
+
+<!-- What hard blockers exist which might prevent further work on this project? -->
```
Team capacity may be a legitimate concern here. At the very least, we would have to pull maintainers from other parts of the project if we plan to do this in Q3-Q4 of this year.

zackkrida commented on documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md:
```diff
@@ -0,0 +1,338 @@
+Since Nuxt 3 specific features will be in a separate branch, if there is a need
+to roll back after merging `nuxt3` branch, we can revert the merge commit.
+
+## Risks
```
Some other risks I can think of:

1. Drift between the `nuxt3` and `main` branches, or difficulty keeping them in sync
2. Encountering an unexpected dependency incompatability that halts the migration, further exacerbating pt.1

zackkrida commented:
This is an excellent breakdown of the sheer mass and complexity of work required to execute this migration. I have to admit I find it terrifying. Particularly the lack of stability and support in many of the ecosystem packages: i18n, nuxt-vitest, storybook (there's a $500 bounty in [this issue](https://github.com/nuxt-community/storybook/issues/330), sentry, and I imagine many more. It seems the only thing that might bring some stability to these dependencies is more time, as many of them are in active development. I worry that the list of changes here represents our known issues, that the list of known issues is quite sizable, and that the list of _unknown_ issues may be even larger. 

Given that this is no longer time sensitive, as #2430 adds Node 18.x support which will be maintained far into 2025, I wonder if an ideal course of action may be to focus on the prerequisite changes, but also any work to audit and clean up the frontend as much as possible beyond what is necessary for Nuxt 3. And to explicitly revisit the Nuxt 3 part of the project several months from now when some of these libraries might have better support.

It's a fine line between delaying the inevitable and waiting for dependencies to make improvements. 

It's also worth considering that two frontend contributors, in fact the reviewers of this proposal, will either be AFK entierely or have reduced capacity at the time of year we might consider implementing this work. 

I could also see the Nuxt 3 steps and branch dragging on for a really long time. I think the workflow to fix these problems could benefit from some pair programming sessions. So often this work involves untangling cryptic dependency errors.

I think I'd like to wait and give this a final approval once any remaining revisions are in place, but the overall approach and breakdown of the tasks makes a lot of sense to me. I find myself really distracted by how gargantuan of an undertaking this will be.

sarayourfriend commented:
> Given that this is no longer time sensitive, as https://github.com/WordPress/openverse/pull/2430 adds Node 18.x support which will be maintained far into 2025, I wonder if an ideal course of action may be to focus on the prerequisite changes, but also any work to audit and clean up the frontend as much as possible beyond what is necessary for Nuxt 3. And to explicitly revisit the Nuxt 3 part of the project several months from now when some of these libraries might have better support.

I strongly support this option now that Node 16 EOL is no longer a factor, especially when [Elasticsearch 17 EOL](https://www.elastic.co/support/eol) is also approaching at the same time we were originally planning to implement this, as @obulat (IIRC, might have been Krystle) pointed out elsewhere.

krysal commented:
@obulat Should this be marked as blocked until #2490 is merged? The project thread says it depends on some TypeScript updates. Could you leave a comment on what is needed to complete this plan?

obulat commented:
Closing this PR until the Additional search views project is done. Then, I'll re-open it with the updated status and plan.
